Allow passed in vertex_ai credentials to be authorized_user type

pselden opened the pull request:
## Title

Allow passed in vertex_ai credentials to be authorized_user type

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🆕 New Feature

## Changes

When using the proxy, we desire to pass in a json string that contains the credentials. In production, using a service account is fine.

However when developing locally, we cannot use the application default credentials that are generated by `gcloud auth application-default login`. When passing as a string, those are currently attempted to be parsed as a service account.

This change enables users to pass that type of credential directly as a string, instead of relying on it loading a file directly on the proxy.

### How to test this for real

`gcloud auth login` -- follow steps to log on
`gcloud auth application-default login` -- follow steps to generate the credentials file

Follow the example code on https://docs.litellm.ai/docs/providers/vertex:

```
from litellm import completion
import json 

file_path = 'path/to/application_default_credentials' # usually $HOME/.config/gcloud/application_default_credentials.json

# Load the JSON file
with open(file_path, 'r') as file:
    vertex_credentials = json.load(file)

# Convert to JSON string
vertex_credentials_json = json.dumps(vertex_credentials)

## COMPLETION CALL 
response = completion(
  model="vertex_ai/gemini-pro",
  messages=[{ "content": "Hello, how are you?","role": "user"}],
  vertex_credentials=vertex_credentials_json
)
```

Prior to this change, the code would attempt to parse it as a service account and error. After this chance, it will correctly load the _user_ credentials and complete the request.

pselden committed:
Allow passed in vertex_ai credentials to be authorized_user type

When using the proxy, we desire to pass in a json object that contains the credentials.
In production, using a service account is fine.
However when developing locally, we cannot use the application default credentials that are generated by
`gcloud auth application-default login` and pass that value to the proxy for authorization.

This change enables users to pass that type of credential directly as a string.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=10899) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=10899) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=10899) it.</sub>

pselden commented:
I'm having trouble getting the tests to pass locally without directly adding google-auth to the project.toml with poetry.

Getting the following:

```
FAILED tests/litellm/llms/vertex_ai/test_vertex_llm_base.py::TestVertexBase::test_authorized_user_credentials[async] - ModuleNotFoundError: No module named 'google.auth'
FAILED tests/litellm/llms/vertex_ai/test_vertex_llm_base.py::TestVertexBase::test_authorized_user_credentials[sync] - ModuleNotFoundError: No module named 'google.auth'
```
When I do add it, it works correctly.

pselden commented on litellm/llms/vertex_ai/vertex_llm_base.py:
```diff
@@ -99,7 +105,7 @@ def load_auth(
                 project_id = creds_project_id
 
-        creds.refresh(Request())  # type: ignore
+        self.refresh_auth(creds)
```
Switched to using this so I could mock it out in a test (the current test suite mocks out this entire method, but I want to actually test this thing)

pselden committed:
Fix comment in test

pselden commented:
Tests passing after doing a `poetry add google-auth` (which also changes hundreds of other things in the lockfile... probably because of a different poetry version)

![image](https://github.com/user-attachments/assets/39c3bdcb-a540-494a-828d-b320fb757a58)

pselden committed:
Merge remote-tracking branch 'upstream/main' into vertex-ai-authorized-user-credentials

pselden committed:
Move auth methods to helpers so they can be mocked in tests

pselden commented:
Still having trouble with dependencies in the tests.

krrishdholakia commented:
Looks like your tests use google.auth - can we mock this? @pselden 
<img width="1164" alt="Screenshot 2025-05-18 at 8 46 27 PM" src="https://github.com/user-attachments/assets/ca9c87f4-65ab-4c64-81fd-2b9a98548550" />

pselden commented:
> Looks like your tests use google.auth - can we mock this? @pselden <img alt="Screenshot 2025-05-18 at 8 46 27 PM" width="1164" src="https://private-user-images.githubusercontent.com/17561003/444963129-ca9c87f4-65ab-4c64-81fd-2b9a98548550.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc2MjcxNDgsIm5iZiI6MTc0NzYyNjg0OCwicGF0aCI6Ii8xNzU2MTAwMy80NDQ5NjMxMjktY2E5Yzg3ZjQtNjVhYi00YzY0LTgxZmQtMmI5YTk4NTQ4NTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTE5VDAzNTQwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzOTZhNTE2ZWRkZTc3YTA3MGExYTBmYjBmMTM0OTRiZWZkYTU1OTc3ODZlMzM2NTI0YWY4ZGZmMzY2NGEzOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.J9pOc2yxlC7f-WGsKyLLyjLNDCacSqKsxP9TAaf_pyY">

The functionality I added depends on  google auth - if I mock it it won’t be testing anything anymore (except the mock itself). I mocked out the actual network  request in refresh_auth.

I’m happy to mock it like the other vertex base tests do if that’s how you handle it, but then it wouldn’t be calling the lines I added at all.

pselden committed:
Merge remote-tracking branch 'upstream/main' into vertex-ai-authorized-user-credentials

pselden commented:
I've pulled all of the auth methods into helpers so they can be mocked independently in the tests, and now they are passing.
